### PR TITLE
feature: Support custom jwt

### PR DESCRIPTION
**Pull request by FrankEssenberger**
Closes SAP/cloud-sdk-backlog#546

Relatest to https://github.com/SAP/cloud-sdk-js/issues/2016.

Add support for JWT that do not have a JKU property e.g. which are not from XSUAA. This is a security relevant change so please do a careful review. Also note that the destination services does a verification so all layers of security we add are just extra and lead to better error messages and error message control on our side.

### CODE REVIEW

File: packages/connectivity/src/scp-cf/destination/destination-accessor-jwks.spec.ts [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+      userJwt
+    );
+
+    const spy = jest.spyOn(jwt, 'verifyJwt');
+    const actual = await getDestination({ ...destFetchOption, jwt: userJwt });
+    expect(spy).toHaveBeenCalledTimes(0);
+    expect(actual).toBeDefined();
+  });
+
+  it('throws an error if no jwks properties are not given for JWT without JKU', async () => {
```
Comment (ZhongpinWang) [inline packages/connectivity/src/scp-cf/destination/destination-accessor-jwks.spec.ts:91]:
```suggestion
  it('throws an error if jwks properties are not given for JWT without JKU', async () => {
```

File: packages/connectivity/src/scp-cf/destination/destination-accessor-jwks.spec.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+    mockOneDestination({ ...oauthMultipleResponse[0] }, serviceJwt, userJwt);
+
+    const spy = jest.spyOn(jwt, 'verifyJwt');
+    const actual = await getDestination({
+      ...destFetchOption,
+      jwt: userJwt,
+      selectionStrategy: alwaysSubscriber
+    });
+    expect(spy).toHaveBeenCalledTimes(1);
+    expect(actual).toBeDefined();
```
Comment (ZhongpinWang) [inline packages/connectivity/src/scp-cf/destination/destination-accessor-jwks.spec.ts:82]: The type of `actual` is `Destination | null`. Therefore, `.not.toBeNull()` should be used instead of `.toBeDefined()` because `expect(null).toBeDefined()` is also valid. I will fix this and check the other occurrences of `.toBeDefined()`

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 3]
```diff
@@ -434,10 +485,14 @@ Possible alternatives for such technical user authentication are BasicAuthentica
...
       throw new Error('Try to get subscriber token but value is undefined.');
     }
-    return (
-      this.subscriberToken.userJwt?.decoded ||
-      this.subscriberToken.serviceJwt.decoded
-    );
+    switch (this.subscriberToken.type) {
+      case 'custom':
+        return this.subscriberToken.userJwt.decoded;
+      case 'iss':
```
Comment (ZhongpinWang) [inline packages/connectivity/src/scp-cf/destination/destination-from-service.ts:491]: [req] I actually wanted to comment on line 483. Since the new `IssToken` interface specifically set `userJwt` to `undefined`, maybe we can remove this line of comment.

File: packages/connectivity/src/scp-cf/jwt.spec.ts [hunk 4]
```diff
@@ -35,6 +42,35 @@ function responseWithPublicKey(key = publicKey) {
 }
 
 describe('jwt', () => {
+  describe('isXsuss', () => {
```
Comment (ZhongpinWang) [inline packages/connectivity/src/scp-cf/jwt.spec.ts:45]:
```suggestion
  describe('isXsuaa', () => {
```

File: packages/connectivity/src/scp-cf/jwt.ts [hunk 5]
```diff
@@ -49,6 +49,24 @@ export function retrieveJwt(req: IncomingMessage): string | undefined {
   }
 }
 
+/**
+ * Checks if the given JWT is from the XSUAA or from a alternative issuer based on the iss property and the uua domain of the XSUAA.
```
Comment (ZhongpinWang) [inline packages/connectivity/src/scp-cf/jwt.ts:53]:
```suggestion
 * Checks if the given JWT is from the XSUAA or from an alternative issuer based on the jku property and the uaa domain of the XSUAA.
```

**Review by ZhongpinWang — COMMENTED** [review]
I found some typos and issues with the naming and consistency.

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 6]
```diff
@@ -156,15 +178,29 @@ class DestinationFromServiceRetriever {
...
+        const encoded = await serviceToken('destination', {
+          ...options,
+          jwt: options.jwt
+        });
+        return {
+          type: 'xsuaa',
+          userJwt,
+          serviceJwt: { encoded, decoded: decodeJwt(encoded) }
+        };
+      }
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination-from-service.ts:199]: did a little clean up

File: packages/connectivity/src/scp-cf/jwt.ts [hunk 7]
```diff
@@ -262,6 +280,14 @@ export interface JwtPair {
   encoded: string;
 }
 
+/**
+ * Build JwtPair from a encoded JWT.
```
Comment (ZhongpinWang) [inline packages/connectivity/src/scp-cf/jwt.ts:284]:
```suggestion
 * Build JwtPair from an encoded JWT.
```

File: .changeset/angry-penguins-flow.md [hunk 8]
```diff
@@ -0,0 +1,7 @@
+---
+'@sap-cloud-sdk/connectivity': minor
+'@sap-cloud-sdk/test-util': minor
```
Comment (jjtang1985) [inline .changeset/angry-penguins-flow.md:3]: I feel this is not touched?

File: .changeset/angry-penguins-flow.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+'@sap-cloud-sdk/integration-tests': minor
```
Comment (jjtang1985) [inline .changeset/angry-penguins-flow.md:4]: We will not release this package, so we should not add this?

File: test-packages/integration-tests/test/cache.spec.ts [hunk 10]
```diff
@@ -122,14 +122,14 @@ describe('CacheDestination & CacheClientCredentialToken', () => {
       useCache: true,
       isolationStrategy: IsolationStrategy.Tenant_User
     });
-    expect(destinationRequestTenantUser).toBeDefined();
+    expect(destinationRequestTenantUser).not.toBeNull();
```
Comment (jjtang1985) [inline test-packages/integration-tests/test/cache.spec.ts:125]: [q] what's the differences?

File: packages/connectivity/src/scp-cf/jwt.ts [hunk 11]
```diff
@@ -49,6 +49,24 @@ export function retrieveJwt(req: IncomingMessage): string | undefined {
   }
 }
 
+/**
+ * Checks if the given JWT is from the XSUAA or from an alternative issuer based on the iss property and the uaa domain of the XSUAA.
+ * @param decodedUserJwt - JWT to be checked.
+ * @returns True if JWT is issued by XSUAA
+ * @internal
+ */
+export function isXsuaaToken(decodedUserJwt: JwtWithPayloadObject): boolean {
```
Comment (jjtang1985) [inline packages/connectivity/src/scp-cf/jwt.ts:58]: Would it be better to share the potential doc of the logic in this function?
So can see potential updates later in newer versions.

File: packages/connectivity/src/scp-cf/jwt.spec.ts [hunk 12]
```diff
@@ -35,6 +42,35 @@ function responseWithPublicKey(key = publicKey) {
...
+      expect(isXsuaaToken(jwt)).toBe(true);
+    });
+
+    it('returns false if jku is missing', () => {
+      const jwt = decodeJwtComplete(signedJwtForVerification({}, undefined));
+      mockServiceBindings();
+      expect(isXsuaaToken(jwt)).toBe(false);
+    });
+
+    it('returns false if jku and uaa are from same domain', () => {
```
Comment (jjtang1985) [inline packages/connectivity/src/scp-cf/jwt.spec.ts:63]:
```suggestion
    it('returns false if jku and uaa are from different domain', () => {
```

File: packages/connectivity/src/scp-cf/destination/destination.ts [hunk 13]
```diff
@@ -380,7 +380,9 @@ const configMapping: Record<string, keyof Destination> = {
   certificates: 'certificates',
   KeyStoreLocation: 'keyStoreName',
   KeyStorePassword: 'keyStorePassword',
-  SystemUser: 'systemUser'
+  SystemUser: 'systemUser',
+  'x_user_token.jwks': 'jwks',
```
Comment (jjtang1985) [inline packages/connectivity/src/scp-cf/destination/destination.ts:384]: This map gets bigger over the time, would it be fine to add some docs about the keys, especially the new ones?

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 14]
```diff
@@ -315,7 +348,7 @@ class DestinationFromServiceRetriever {
     const exchangeTenant = this.getExchangeTenant(destination);
     const clientGrant = await serviceToken('destination', {
       jwt:
-        origin === 'subscriber'
+        origin === 'subscriber' && this.subscriberToken?.type !== 'custom'
```
Comment (jjtang1985) [inline packages/connectivity/src/scp-cf/destination/destination-from-service.ts:351]: Is this covered by some tests from a test driven perspective?

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 15]
```diff
@@ -346,16 +379,31 @@ Possible alternatives for such technical user authentication are BasicAuthentica
...
       );
     }
     // This covers OAuth to user dependend auth flows https://help.sap.com/viewer/cca91383641e40ffbe03bdc78f00f681/Cloud/en-US/39d42654093e4f8db20398a06f7eab2b.html and https://api.sap.com/api/SAP_CP_CF_Connectivity_Destination/resource
     // Which is the same for: OAuth2UserTokenExchange, OAuth2JWTBearer and OAuth2SAMLBearerAssertion
 
-    // Case 1 Destination in provider and jwt issued for provider account -> not extra x-user-token header needed
-    if (this.isProviderAndSubscriberSameTenant()) {
+    // If user JWT is not XSUAA enforce the JWKS properties are there - destination service would do that as wll. https://help.sap.com/docs/CP_CONNECTIVITY/cca91383641e40ffbe03bdc78f00f681/d81e1683bd434823abf3ceefc4ff157f.html
+    if (this.subscriberToken.type === 'custom') {
+      if (!destination.jwks && !destination.jwksUri) {
```
Comment (jjtang1985) [inline packages/connectivity/src/scp-cf/destination/destination-from-service.ts:392]: [nth] we can wrap this if condition to some `validateCustomJWT` functions for better readability.

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 16]
```diff
@@ -346,16 +379,31 @@ Possible alternatives for such technical user authentication are BasicAuthentica
...
+        throw new Error(
+          'Failed to verify the JWT with no JKU! Destination must have `x_user_token.jwks` or `x_user_token.jwks_uri` property.'
+        );
+      }
+    }
+
+    // Case 1 Destination in provider and jwt issued for provider account But not custom JWT given-> not extra x-user-token header needed
+    if (
```
Comment (jjtang1985) [inline packages/connectivity/src/scp-cf/destination/destination-from-service.ts:400]: [q] Are the case 1 and 2 covered by some unit tests, as they are now more complicated?

**Review by jjtang1985 — COMMENTED** [review]
I left some suggestions and questions.
For me, the code looks more complicated than before (was a bit complicated already), I hope the new logics are covered by the current unit tests, so we avoid similar edge cases like the jwt vulnerability before.
If you like, you can also use the test coverage report and check the differences.
On the other hand, having auth-flow test is the first step, do we consider some e2e tests for this?

File: .changeset/angry-penguins-flow.md (hunk 8 above)
Comment (FrankEssenberger) [reply .changeset/angry-penguins-flow.md:3]: I confused the `test-util` from the test-resources with the test-util we ship. Removed

File: .changeset/angry-penguins-flow.md (hunk 9 above)
Comment (FrankEssenberger) [reply .changeset/angry-penguins-flow.md:4]: also removed

File: packages/connectivity/src/scp-cf/destination/destination.ts (hunk 13 above)
Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination.ts:384]: added commnet for the new ones

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 17]
```diff
@@ -53,11 +60,27 @@ interface DestinationSearchResult {
...
+interface XsuaaToken {
+  type: 'xsuaa';
+  userJwt: JwtPair;
   serviceJwt: JwtPair;
 }
 
+interface CustomToken {
+  type: 'custom';
+  userJwt: JwtPair;
+  serviceJwt: undefined;
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination-from-service.ts:81]: added some docs

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts (hunk 14 above)
Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination-from-service.ts:351]: yes, I inverted the logic so I ask for origin `provider` which covers the custom case as well

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts (hunk 15 above)
Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination-from-service.ts:392]: added some helper - sligthly different to your suggestion called `checkXYZ` also throwing the exception.

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts (hunk 16 above)
Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination-from-service.ts:400]: yes there are multiple test, I also changed the logic because custom means that the provider service token must be used so it is not really more complicated.

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 18]
```diff
@@ -147,6 +192,20 @@ class DestinationFromServiceRetriever {
...
+      throw new Error(
+        'Failed to verify the JWT with no JKU! Destination must have `x_user_token.jwks` or `x_user_token.jwks_uri` property.'
+      );
+    }
+  }
+
+  private static isUserJwt(
+    token: SubscriberTokens | undefined
+  ): token is CustomToken | XsuaaToken {
+    return !!token && token.type !== 'iss';
```
Comment (marikaner) [inline packages/connectivity/src/scp-cf/destination/destination-from-service.ts:206]: [req] You can shorten this.
```suggestion
    return token?.type !== 'iss';
```

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 19]
```diff
@@ -156,29 +215,39 @@ class DestinationFromServiceRetriever {
...
         ...options,
-        jwt: payload
+        jwt: { iss: options.iss }
       });
-      const clientCertJwt = { encoded, decoded: decodeJwt(encoded) };
-      return { serviceJwt: clientCertJwt };
+      return {
+        serviceJwt: getJwtPair(serviceJwtEncoded),
+        type: 'iss',
+        userJwt: undefined
```
Comment (marikaner) [inline packages/connectivity/src/scp-cf/destination/destination-from-service.ts:249]: [pp] Totally irrelevant, I just like it more:

```suggestion
        type: 'iss',
        userJwt: undefined,
        serviceJwt: getJwtPair(serviceJwtEncoded)
```

File: packages/connectivity/src/scp-cf/destination/destination-from-service.ts [hunk 20]
```diff
@@ -53,11 +60,49 @@ interface DestinationSearchResult {
...
+type SubscriberTokens = IssToken | XsuaaToken | CustomToken;
+
+/**
+ * User provided a dummy token with the `iss` property.
+ * This is used if a different tenant than the provider ones should be accessed but no user related login (JWT) for this tenant is available.
+ * Therefore, the userJwt is undefined and only a destination service token has been issued.
+ */
+interface IssToken {
+  type: 'iss';
+  userJwt: undefined;
```
[the remark this answers is not in the document]
Comment (FrankEssenberger) [reply packages/connectivity/src/scp-cf/destination/destination-from-service.ts:78]: yes internal. The user sets the iss property on the `destinationFetchOptions` if this is done it leads to that token internally.

File: packages/connectivity/src/scp-cf/jwt.ts (hunk 11 above)
Comment (jjtang1985) [reply packages/connectivity/src/scp-cf/jwt.ts:58]: discussed.
The jku is a standard key for saving domain, so you can check xsuaa url.

**Review by jjtang1985 — APPROVED** [review]
LGTM
